### ISSUE TITLE
Add adjustable seat option

### ISSUE THREAD

**Issue by wlbragg** [c172p-team/c172p#1199]
Even though we have the ability to change the view point by other means, I would like an easy way to make this a permanent change in the context of adjusting my seat. I was searching through a few different 172p POH manuals (80's) and it is clear there is at least an option of a hand cranked vertical adjustment seat.
I always adjust my view position so I can see the nose of the aircraft, it is a minor thing but necessary every time. I look at it as being more important than setting the visor or air vent so I'd like to program it as an option.

If I work on this, or anyone else even, I think I would make a new seat mount with the appropriate hardware that could be clicked in as well as a GUI option.

**Comment by legoboyvdlp** [comment]
It seems only the six-way seats can have their height adjusted:

![image](https://user-images.githubusercontent.com/10434911/45170669-5328f900-b1f8-11e8-8b95-954ab90b6a82.png)

![image](https://user-images.githubusercontent.com/10434911/45170682-5ae89d80-b1f8-11e8-93ab-2ca9acce0b30.png)

![image](https://user-images.githubusercontent.com/10434911/45170694-61771500-b1f8-11e8-8a92-38f3dda26966.png)


I take it we can assume we have six-way seats.

**Comment by gilbertohasnofb** [comment]
I think this could be a nice feature for the next release, and should also be easy to implement.

**Comment by legoboyvdlp** [comment]
Other aircraft e.g. f16, viggen have nice implementations of moving views to specific locations. I think it's mostly angle (looking down at displays) but the same technique can be used. I think it's using interpolate() functions in nasal.

**Comment by wlbragg** [comment]
I think this is the seat frame I want to model. Looks like simple mechanisms to move for/aft and raise/lower. The raise/lower looks like it also moves for/aft slightly.

Everyone OK with this, it was in a 172m? If anyone can verify it fits a 172p I would appreciate it. I don't see why it wouldn't .
That leads me to another question, should this be an option (choice of two seat frames) or just make the only seat option adjustable?

**Comment by gilbertohasnofb** [comment]
@wlbragg I think precise measurements might not help us too much given that our interior was not build using precise measurements. For instance, when we raised the panel, we did it so that it looked good to our eyes and similar to photographs, but the height, depth, angles, etc., are not precise at all. As for giving the user a choice of two seat frames, I feel that is a bit too much. For one, people's individual heights aren't simulated, so such real life procedures seem very unnecessary to me. I thought that this issue here would be solved by simply giving the users a slider which would let them tweak the height of the pilot and copilot seat using some arbitrary references we judge good, basically an easy way of doing the same as middle clicking with the mouse but also imposing a reasonable range to move up and down. From the 3D model perspective, I don't think we need to do anything, I think that literally raising the seat and the pilot/copilot model would also be an overkill too (the human models are all out of proportion in FG, anyway). My 2 cents of how to approach this.

**Comment by wkitty42** [comment]
On 09/28/2018 05:10 PM, Gilberto Agostinho wrote:
> I thought that this issue here would be solved by simply giving the users a
> slider which would let them tweak the height of the pilot and copilot seat
> using some arbitrary references we judge good, basically an easy way of doing
> the same as middle clicking with the mouse but also imposing a reasonable
> range to move up and down. From the 3D model perspective, I don't think we
> need to do anything, I think that literally raising the seat and the
> pilot/copilot model would also be an overkill too (the human models are all
> out of proportion in FG, anyway). My 2 cents of how to approach this.

i agree, gilberto... making it easy for the user to raise or lower their eye 
height is one thing but i don't think we really need to model and animate a 
while seat structure... wlbragg or anyone else can model the seat, sure, but i 
don't think it needs to go much further than that... i've already seen a few 
comments in the forum about the c172p getting a little heavy for some apparently 
low end or older systems...

**Comment by callahanp** [comment]
I just happened to have the parts manual open when I saw this post.

The parts manual shows the "infinitely adjustable front seat installation"
on pages 260 & 266.

Infinitely? That sounds a bit optimistic!

Version 1 of it indicates serials 17263459 through 17276079.  Version 2 is
17276080 on.
So at least it was available from 1975 on.  The simpler seats covered
similar ranges.  It appears to me that both were available from 1975
through 1984.  I don't know if one or the other was standard in any year.
in 1979 the back adjust changed from a separate lever to a gas
cylinder/spring with a lock operated by the front/back lever.

The wide lever under the front is for rolling forward and back. There's a
back adjust lever in there too but I think that disappeared after 1979 and
they used a gas cylinder/spring  to control the back that probably locked
in position when you set the forward and back position.

I haven't checked the POH on how the hand cranks operated or which does
what to which part of the seat.

-pat

**Comment by wlbragg** [comment]
>the human models are all out of proportion in FG, anyway

I wasn't even considering this problem at all and I agree, I don't want to get into having to account for human models in any way, shape or form because we move the seat physically.

>I thought that this issue here would be solved by simply giving the users a slider which would let them tweak the height of the pilot and copilot seat using some arbitrary references we judge good, basically an easy way of doing the same as middle clicking with the mouse but also imposing a reasonable range to move up and down

Sounds reasonable to me and so long as it makes this choice persistent, I am OK with this simple concept.

>From the 3D model perspective, I don't think we need to do anything,

Need to, no. However I  never liked the existing seat mounts all that much and I started making a mount more like the one in the picture I posted. I still may finish it. It would give us a fore/aft bar and a height knob that could be made to be hot so that when moused on could do the same thing the slider would do and that would be only to change the eye perspective but not change the seat model mesh position in any way.

>I think that literally raising the seat and the pilot/copilot model would also be an overkill

That was my original intent, but I agree, we don't need to do this and with the human model involved, I don't want to go down that road. Just a way to position the view in a persistent manner is good enough for me. If we get a better seat mount out of the deal then hopefully everyone would be OK with that.

>If anyone can verify it fits a 172p I would appreciate it. I don't see why it wouldn't.
That leads me to another question, should this be an option (choice of two seat frames) or just make the only seat option adjustable?

@callahanp thank you for the information.

My first concern here was making sure if I did model a new seat mount, that it was appropriate for our model of aircraft.
As far as a choice. I was thinking along those same lines, that we make sure we have a realistic version available, even if the optional adjustable seat mount wasn't standard, thus where a choice came into being.
I am more than happy with no choice, only the optional 6-way seat frame is fine with me. I feel if we are going to provide the ability to change our eye view and that change is regarded as simulating the ability to change the seat position (even if not physically showing the seat movement) then we should be OK with a seat mount that fits that ability, even if we don't allow it to physically move. So long as we agree that is a realistic even though an optional setup for this model of c172p.
I don't think it would be appropriate to have a non-adjustable seat mount and the ability to change the eye view in a persistent manner such as you would get from an adjustable seat. That doesn't seem realistic to me.
So if we are going to give the ability to make your eye view persistent as if we are simulating the ability to move the seat then I think we should model a 6-way seat frame, even if we don't model the animation to go with it.
I don't mind modeling  a new representative seat frame and it will look better than the existing one. I also can assure anyone that is concerned about the overhead of such a model that it won't be any heavier than what we already have. The mounts we already have are not optimized at all. As far as overkill, animating the seat model is probably overkill and especially with the human model involved in any animation a pain we don't need to deal with. But simply replacing the model with the appropriate 6-way seat model to represent the new capability we are implementing, I don't see it as overkill at all, simply an upgrade like any other we have done, from vertical stabilizer to sun visors to strut attachment points to disk brakes.

Anyway, I am the one willing to do the modeling work, and I don't see any down side to it, so I would hope everyone would be OK with that.

My only concern all along is keeping it realistic.

**Comment by callahanp** [comment]
where are you sourcing your seat design from?  do you have a copy of the
parts catalog?

**Comment by wlbragg** [comment]
@callahanp what a great idea. I never once thought of searching for a parts catalog only user and op manual. The parts catalog is perfect source for reference.
My source was a photo of a seat being installed in a 172m, however after looking at the parts catalog it looks identical (or close enough) to the above mentioned "infinitely". Good thing I looked thought as I was going to install it upside down. Our back seat frame does not at all resemble the real life frame, I can fix that while I am at it.
The front seats have a plastic cowling around at 
The hand crank knob on the seat I think I am going to model moves the seat up and down. But the action of the moving mechanism looks like it also moves it slightly fore and aft in the process. It looks like it pivots a bar around an attachment point at one end from something less than horizontal to vertical, the length of the bar appears to be range of height capability.

**Comment by wlbragg** [comment]
@callahanp I agree with your analysis of the seat mechanisms.
The "infinitively" seat as best I can tell has three adjustments...

1) Elevation (slight fore/aft motion as well) bellcrank lower left front.

2) Lever bar or lever knob for rolling forward and aft.

3) Either an electrical cylinder controlled seat back position or a bellcrank upper center or a lever at the seat back intersection like on a car. I've seen all three configurations.

Note, the seat back position bellcrank or knob is on the upper cushion frame VS the lower seat mount frame.
The elevation bellcrank and fore/aft lever or knob is on the lover seat mount frame.

Obviously we will offer only height and maybe some limited fore and aft view change.

**Comment by wlbragg** [comment]
Here is the final frame for the front seats. They've been done for awhile, i kind of got distracted and haven't gone any further. @gilbertohasnofb I think part of the reason I stopped working on it is the realization we need new textures for it and thus light map adjustments. My thought were to use the same space we use now for the frames to put new textures down. I am going to do the back frame as well. It will be a very simple tube frame like the real life one is. It is one big frame VS the way we do front seats. Our back seat frame is currently not realistic at all.

**Comment by gilbertohasnofb** [comment]
That looks really good!!

> I think part of the reason I stopped working on it is the realization we need new textures for it and thus light map adjustments

Well, textures aren't a problem. I can simply replace the old ones with a more suitable grey metal to be applied to all mounts. Also, I will make it more uniform as in your screenshot and less corroded, that old texture doesn't look that good IMO. About the lightmaps, I don't seem to understand what you mean. Surely there is barely any light at all hitting these points, and anyway I can easily adjust the lightmap of `interior.png` as to make them dark. Or am I misunderstanding you somehow?

**Comment by gilbertohasnofb** [comment]
For instance, I can modify the current texture:

![image](https://user-images.githubusercontent.com/5700795/46760460-1013e700-ccca-11e8-8d1f-12cd63d0bbf5.png)

into something like this:

![image](https://user-images.githubusercontent.com/5700795/46760474-13a76e00-ccca-11e8-9643-1cfebe27b9be.png)

**Comment by wlbragg** [comment]
>Surely there is barely any light at all hitting these points, and anyway I can easily adjust the lightmap of interior.png as to make them dark. Or am I misunderstanding you somehow?

No, just dealing with them at all. I think I have had enough light mapping to last awhile. After I did these I had to redo the set for the shuttle because I added a single object to the cabin. They are very complex and it was just a re-learning curve.
But you right, there is so little light reaching that area it really is nothing to do by hand.

Could you add a couple "rod" and end cap textures to the top part of the new texture. 

![kill](https://user-images.githubusercontent.com/11069680/46761703-d11a6d00-cc9a-11e8-9bd0-ed2066a51622.jpg)

You can safely use up to a 1/4 of it I think. It's just a little extra that will really spice up the seat mounts. Those three objects are ones I am using in the AirCrane and are GPL.

1) Maybe 2 of each of the 3 textures will suffice. Whatever fits there nicely will do (even one of each). 
2) Also maybe a black or dark "plastic" knob for the crank handle ends. 
3) Even a couple of the "Phillips head screws" if you feel like it.

I'll be using some of this for the backseat frame as well.

I'm not too concerned about overlapping uv mappings down where these objects are located because of the lack of detailed light that will reach the area. Not to mention no one is going to be looking that close.

**Comment by gilbertohasnofb** [comment]
Cool, I will try to get this done soon.

**Comment by wlbragg** [comment]
@gilbertohasnofb  be sure to use the frost and fog branch to change the texture because I am using it for the seat mounts as I also have mesh changes for the windows (at least their settings in the .ac.
I won't be making a new branch for the seat mounts.

**Comment by gilbertohasnofb** [comment]
> I won't be making a new branch for the seat mounts.

All right, but why did you use the same branch? I think it's really good to keep unrelated things separated. It also mean that we can only merge that PR when both features are implemented.

**Comment by wlbragg** [comment]
Because it is impossible to do mesh edits in two different branches and keep it straight. I had to change some properties in the blender and .ac to do the fog/frost and that would be lost if I also tried to make a branch with blender changes for the seat mounts.

I actually started to do that and realized I was asking for trouble if I even tried. Because one binary will overwrite the other. You could probably do it with the .ac but not the blend. Even the .ac could potentially be problematic.

I guess what I should have done in retrospect is to finish one before starting on the other.

**Comment by gilbertohasnofb** [comment]
Oh, I see, I forgot that the other branch also had mesh modifications. Anyway, it's all good, I will make the texture soon.

**Comment by gilbertohasnofb** [comment]
@wlbragg Done. I pushed only for the 4X-CHV since we might still need to modify this texture. But please make it livery dependent for the sake of consistency, even though I plan to use the same mount texture for all liveries. Let me know if the new texture works for you.

**Comment by wlbragg** [comment]
@gilbertohasnofb perfect, thank you!

**Comment by wlbragg** [comment]
![fgfs-screen-017](https://user-images.githubusercontent.com/11069680/46779546-f11f5000-ccdd-11e8-855e-1ee08bc3c754.png)

**Comment by wlbragg** [comment]
![fgfs-screen-020](https://user-images.githubusercontent.com/11069680/46781139-cc2edb00-cce5-11e8-8650-f01825d0d05b.png)

**Comment by wlbragg** [comment]
I added some details like bolts anchoring it down. I also noticed in the parts manual all the back seats while having a bench seat have individual seat backs that can be adjusted individually.
![fgfs-screen-023](https://user-images.githubusercontent.com/11069680/46782272-37c77700-cceb-11e8-8441-929a53b42206.png)

**Comment by legoboyvdlp** [comment]
Looks great!

**Comment by gilbertohasnofb** [comment]
@wlbragg looks amazing. If you think you won't need any more texture changes then I will push those modifications to all other liveries. I will wait for you to give me green light to do that.

**Comment by wlbragg** [comment]
@gilbertohasnofb can you lighten the dark grey part of the texture? Maybe a bit lighter metal shade. It's not mandatory but I think it'll look better.

**Comment by gilbertohasnofb** [comment]
@wlbragg done, let me know if this works for you. I also pushed a fixed to this https://forum.flightgear.org/viewtopic.php?f=4&t=25157&sid=228101e9b57ceb9bb08360e0a77eb1af&p=337417#p337417

The user correctly noticed that the panel of N73429 had the wrong call sign.

**Comment by wlbragg** [comment]
@gilbertohasnofb perfect, thanks. Feel free to push it to the rest of the liveries if you haven't done so already.

**Comment by gilbertohasnofb** [comment]
Great, so I will do it tonight.

**Comment by gilbertohasnofb** [comment]
Done!

**Comment by wlbragg** [comment]
I'm ready to finish this issue up but still have a couple questions.
So far I have the seat options as follows...

u = up
U = down
b = forward
B = backward

That is the only movement options currently.
I assume we weren't going to give a GUI option?
I would like to also have a persistent position for view pitch. For example, our default view is looking a bit down on the panel. I would like to use the seat recline option to actually adjust the pitch view and also make it persistent.
So I would add another set of key controls.
I was starting to think about what keys to use for these options, some random thoughts...

Unused in the c172p but may be pertinent elsewhere
u/U = newly assigned to seat up/down
b/B = newly assigned to seat forward/back
w/W
c/C
t/T
p
F
d

Used but could be rearranged
f = flashlight
D = cabin light (D was for dome light)

I'm thinking along the lines
f = seat forward
b = seat back
u = seat up
d = seat down
U =seat  pitch up
D = seat pitch down

c = cabin light
C = flashlight or t = flashlight (for torch)

Just thinking out loud, suggestions anyone?

I can push this as soon as we decide on these last questions and key assignments.

**Comment by wlbragg** [comment]
Reminder to myself, if we agree to a view pitch adjustment then I still need to add the lever for the seat recline or back adjustment.

**Comment by gilbertohasnofb** [comment]
> Adjusting your seat using the key codes currently is automatically persistent. 

That's excellent, this way users can set up their seat position and never have to touch it again!

> Do we need to make the persistence choice a GUI option?

Possibly yes? Also, a GUI dialog for the seat position could also feature a button to reset any changes to the default position.

**Comment by gilbertohasnofb** [comment]
@wlbragg in what PR are you working on this? I would love to test it later today.

**Comment by wlbragg** [comment]
This is in the fog/frost branch. Q currently resets to the defaults.

**Comment by onox** [comment]
Did you change the keybinding of Q? It's a key used by several models to move the camera back to its original position and rotation.

**Comment by legoboyvdlp** [comment]
Hopefully the keybindings of the c172l wouldn't affect the keybindings of other aircraft?

**Comment by wlbragg** [comment]
No, it's still view reset.

**Comment by gilbertohasnofb** [comment]
@wlbragg Really cool feature and I love that the setting is persistent between sessions. The only pity is that the seat pitch with `U` and `D` is reset when you use a right mouse click in view mode which resets the view and look forward, but I don't think we have a way around that.

**Comment by legoboyvdlp** [comment]
I think that issue was brought up recently on the mailing list and behaviour might be configurable in the view dialog in the next update?

**Comment by gilbertohasnofb** [comment]
I must have missed that, lego. I have been struggling to keep up with my emails lately and I have not been following the devel list for the past weeks. But good to know the issue has been brought up already.

**Comment by wlbragg** [comment]
Anyone know why the z axis is in the pilot_view function VS the reset_view function" Or better yet, why is there a reset_view function at all. It could all be done in the pilot_fies function. There is no distinction between z in pilot_view and x and y in reset_view, they do the exact same thing in the exact same way, why are they not together?

~~~
var reset_view = func (dt) {
    var fow = getprop("sim/current-view/config/default-field-of-view-deg");
    var heading_deg = getprop("sim/current-view/config/heading-offset-deg");
    var pitch_deg = getprop("sim/current-view/config/pitch-offset-deg");
    var roll_deg = getprop("sim/current-view/config/roll-offset-deg");

    var current_heading_deg = getprop("sim/current-view/heading-offset-deg");

    if (current_heading_deg > 180.0) heading_deg += 360.0;

    interpolate("sim/current-view/field-of-view", fow, dt);
    interpolate("sim/current-view/heading-offset-deg", heading_deg, dt);
    interpolate("sim/current-view/pitch-offset-deg", pitch_deg, dt);
    interpolate("sim/current-view/roll-offset-deg", roll_deg, dt);

    if (getprop("sim/current-view/internal")) {
        var index = getprop("sim/current-view/view-number");

        var x_offset = views[index].getValue("config/x-offset-m");
        var y_offset = views[index].getValue("config/y-offset-m");

        interpolate("sim/current-view/x-offset-m", x_offset, dt);
        interpolate("sim/current-view/y-offset-m", y_offset, dt);

        setprop("/controls/current-view/pitch-offset-deg", pitch_deg);
    }
};

var pilot_view = func {
    reset_view(dt);

    if (getprop("sim/current-view/internal")) {
        var index = getprop("sim/current-view/view-number");
        var z_offset = views[index].getValue("config/z-offset-m");

        interpolate("sim/current-view/z-offset-m", z_offset, dt);
    }
};
~~~

Also, after reviewing this some more, I am using the wrong properties to cache. I originally had new properties tied to the original properties and did away with that. but I think I can work this in such a way I may be able to get around the reset changing the pitch. Also i should be able to add [q] to reset the view to the saved adjusted seat view.
So...
Q = reset to default view
q = reset to cached adjusted seat view

**Comment by wlbragg** [comment]
OK, please review the changes, I think you'll like it. Feedback, comments and concerns are welcome.
I added a quick key [q] to quickly reset the view to the user defined seat position. So now there are two choices.
[q] = user defined seat view
[Q] = default view

Do we need a GUI choice to reset the user defined seat view back to the default view or simply let them readjust the seat back to that position in the same way they adjusted it in the first place, with the keys?
Do we need or want a GUI choice to choose default view or user defined?
This may affect the answer to the first question, such as it would no longer be necessary to reset the user defined seat view as you could switch to the default view. Also keep in mind, both of these views are now available with Q or q.

**Comment by wlbragg** [comment]
I haven't done anything with the FOV or the mouse reset (left click). 
I was thinking the shift + left click could reset you to the user defined view?
Do we want to include a persistent setting, either keyboard set or GUI set, for FOV?

**Comment by wlbragg** [comment]
The mouse click one is a bit complicated as it is a standard across FG. I'm not sure we should overwrite this default. People that use it are used to it taking them back to the default view the sim starts in, although we are adding a new dimension of allowing them to set their own view.  Adding a simple shift to let you quickly get back to your user defined view seems simple enough to me.
>one more layer of complication for something supposedly extremely simple.

I don't see it as complicated. I see it as the same difference as adding the q for user defined when we already had Q for the default view.

**Comment by onox** [comment]
> There is no distinction between z in pilot_view and x and y in reset_view, they do the exact same thing in the exact same way, why are they not together?

Are you sure? Z should handle zooming. I think that's why I separated it into two functions, `reset_view` doesn't reset the zooming, just position and orientation.

**Comment by wlbragg** [comment]
Pretty sure, I recoded the two functions into one and named it reset_view VS pilot_view.

pilot_view called reset_view which adjusted  x, y (position) and pitch, roll, yaw and FOV ( orientation) then handled z (position) the same way it handled x and y in the call to reset.
I renamed it because it all deals with reseting the view in all views VS anything specific to pilot view. x, y, and z all are positional view changes.
z = forward and back
y = up and down
x = left and right
My point was there is no need to separate z out from the group as it is called in the same way and time as the rest including roll, pitch, yaw and fov.

This was x and y...
~~~
if (getprop("sim/current-view/internal")) {
        var index = getprop("sim/current-view/view-number");

        var x_offset = views[index].getValue("config/x-offset-m");
        var y_offset = views[index].getValue("config/y-offset-m");

        interpolate("sim/current-view/x-offset-m", x_offset, dt);
        interpolate("sim/current-view/y-offset-m", y_offset, dt);
    }
~~~
This was is z..
~~~

if (getprop("sim/current-view/internal")) {
        var index = getprop("sim/current-view/view-number");

        var z_offset = views[index].getValue("config/z-offset-m");

        interpolate("sim/current-view/z-offset-m", z_offset, dt);
    }
~~~
These are identical in nature and all this is doing is adding and extra if and getting the same property twice for no apparent reason.

The way the entire thing worked was the function (pilot) called the function (reset) to handle x and y, along with pitch, roll, yaw and fov then (pilot) finished with handled z, duplicating the same steps took in reset.

**Comment by wlbragg** [comment]
>reset_view doesn't reset the zooming, just position and orientation.

Even if z is zoom, which I am not sure I understand the meaning here as it is also in this instance now being used for a seat position change of forward and back, not zoom.  I guess I would ask why it would not be part of the reset function. Especially when if you call reset you change it just like the rest, only you duplicate effort and call it function pilot_view. Yet it changes in all the vies, not only pilot view.

**Comment by wlbragg** [comment]
b = seat back 98

Is not going to work, b is used for brakes. I can't believe no one else noticed this yet.

**Comment by gilbertohasnofb** [comment]
Yeah, I didn't realize it since I was always testing the camera while the plane was parked. Do you have another suggestion of what key to use instead?

**Comment by legoboyvdlp** [comment]
is ctrl b an option? Another alternative is maybe k for bacK?

**Comment by gilbertohasnofb** [comment]
If we decide to use `Ctrl` (or any other modifiers) then we must use it for _all_ other seat adjustment shortcuts to keep things consistent. An example could be:

`modifier + PageUp`: move seat up
`modifier + PageDown`: move seat down
`modifier + up`: move seat forward
`modifier + down`: move seat backwards
`modifier + right`: pitch seat angle forwards
`modifier + left`: pitch seat angle backwards

The question would be to find a reasonable modifier or combination of modifiers. Also, `Ctrl + Alt` should be avoided as, on Linux, it's used for a bunch of default keyboard shortcuts.

We could then keep `q` and `Q` for user set and default views, respectively.

Just an idea, of course...

**Comment by wkitty42** [comment]
On 11/27/18 4:42 PM, Gilberto Agostinho wrote:
> Yeah, I didn't realize it since I was always testing the camera while the
> plane was parked. Do you have another suggestion of what key to use instead?

is 'r' available? i'm thinking of 'rear' or 'rearward'...

**Comment by wlbragg** [comment]
The reason this kind of stalled was due to discussion on the dev list and the possibility of extending it in the view code in fgdata.

I/we really need to decide if we want to...

1) Wait for someone else to extend "view" manager in fgdata (I was of the understanding others have brought it up and "may" be working on it).
2) Extend it ourselves in fgdata, not something I am keen on having to learn, but would consider.
3) Go ahead and implement our version local and deal with it if it is ever extended in fgdata.

Comments?

**Comment by gilbertohasnofb** [comment]
@wlbragg What is the likelihood of this being picked up by someone in the dev-list? If you are reasonable positive this might happen in the near future then option 1 is best, else option 3 sounds good to me.

**Comment by wlbragg** [comment]
If I get help from James I could probably get this done in fgdata. I think it is a better way to go if it can get done. I'll take another look at the fgdata side and if I can decipher it I'll try to  do it there.
If not then I should ask the question on the list if it is going to be picked up by anyone anytime soon, if not then we finish this and move on.
I want the ability, I'm not getting stuck on the method!

**Comment by wlbragg** [comment]
I'm definitely finishing the surface friction issue before I start this. Hopefully that can happen in the next week or two.

**Comment by wlbragg** [comment]
@legoboyvdlp  if we finish the defrost issue I am of the opinion we go ahead and also finish the seat view and not count on anyone else at core or even me to do the core code replacement for what I have already done here in this work. In other words I think we should just go with option 3.
>Go ahead and implement our version local and deal with it if it is ever extended in fgdata.

To do that I think all that is missing for the seats is to decide on the modifier. 

I think @gilbertohasnofb idea of
~~~
modifier + PageUp: move seat up
modifier + PageDown: move seat down
modifier + up: move seat forward
modifier + down: move seat backwards
modifier + right: pitch seat angle forwards
modifier + left: pitch seat angle backwards
~~~
and keeping the qQ for quick default and quick user defined are good choices.
I'm OK with using CTRL for the modifier unless @gilbertohasnofb meant
>Also, Ctrl + Alt should be avoided as, on Linux, it's used for a bunch of default keyboard shortcuts.

means using either one of those exclusively?

**Comment by gilbertohasnofb** [comment]
> means using either one of those exclusively?

No, I meant combined. Ctrl plus Alt should be avoided on Linux, but Ctrl alone or Alt alone are fine. We just need to make sure that whatever option we pick does not conflict with default shortcuts for FG. Possibly Alt will be a safer option than Ctrl, as I have the ~totally unscientific opinion~ impression that it is far more seldomly used

**Comment by wlbragg** [comment]
@gilbertohasnofb 
Yes, your right about alt being used less than ctrl. But for some reason I can't get either alt or ctrl modifier to work. See anything wrong here?
~~~
    <key n="361">
      <name>PageDown</name>
      <desc>Decrease throttle or autopilot autothrottle</desc>
      <repeatable type="bool">true</repeatable>
      <binding>
       <command>nasal</command>
       <script>
         controls.incThrottle(-0.01, -1.0)
       </script>
      </binding>
      <mod-shift>
       <desc>Look back right</desc>
       <binding>
        <command>property-assign</command>
        <property>/sim/current-view/goal-heading-offset-deg</property>
        <property>/sim/view/config/back-right-direction-deg</property>
       </binding>
      </mod-shift>
      <mod-alt>
        <desc>Seat Down</desc>
        <binding>
          <condition>
            <equals>
              <property>sim/current-view/view-number</property>
              <value>0</value>
            </equals>
          </condition>
          <command>property-adjust</command>
          <property>sim/current-view/user/y-offset-m</property>
          <step type="double">-0.005</step>
          <min type="double">0.15</min>
          <max type="double">0.25</max>
        </binding>
      </mod-alt>
    </key>
~~~

**Comment by wlbragg** [comment]
Everything with the exception of the mod-alt portion is straight from fgdata keyboard.xml

**Comment by wlbragg** [comment]
I tried it without the view = 0 condition as well.

**Comment by wlbragg** [comment]
References:
http://wiki.flightgear.org/Howto:Reassign_keyboard_bindings#Complete_layouts
http://wiki.flightgear.org/Keyboard_function_priority_list

**Comment by gilbertohasnofb** [comment]
I don't see any issues. Perhaps you can try swapping `mod-alt` for `mod-ctrl` just for testing if that works. If that doesn't work either, you can try deleting the view part of the code and using `mod-shift`, again just for the sake of testing.

**Comment by gilbertohasnofb** [comment]
I also noticed that the `keyboard.xml` file on FGDATA does not use a single shortcut with Alt, so could this be a bug that went unnoticed and the alt modifier is broken?

**Comment by wlbragg** [comment]
I started with ctrl and when that didn't work I tried alt. I can try the shift but I am pretty sure it will work. The view moving property-adjust code worked fine with normal key assignments.
Frustrating, I really didn't want to take this to dev mail list as I don't want to open the can of worms about doing it in the source.

**Comment by gilbertohasnofb** [comment]
Have you seen this on the `keyboard.xml` file? They are not using `mod-ctrl` but rather `Ctrl-<key>` on name:

```
 <key n="2">
  <name>Ctrl-B</name>
  <desc>Toggle speedbrake</desc>
  <binding>
   <command>property-toggle</command>
   <property>/controls/flight/speedbrake</property>
  </binding>
 </key>
```

also, this is an interesting note on line 134:
```
 <key n="10"> <!-- OSG -->
  <name>Enter</name>
  <desc>Move rudder right</desc>
  <repeatable type="bool">true</repeatable>
  <binding>
   <command>property-adjust</command>
   <property>/controls/flight/rudder</property>
   <step type="double">0.05</step>
  </binding>

  <!-- Ctrl-M generates Enter, but we want to show the map in that case -->
  <mod-ctrl>
      <binding>
    	<command>dialog-show</command>
    	<dialog-name>map</dialog-name>
      </binding>
  </mod-ctrl>
 </key>
```

**Comment by wlbragg** [comment]
I'm an idiot. I busted my own code. I made a stupid mistake and the key codes were mostly working.
I've a bit more to do and then we can test it.

**Comment by wlbragg** [comment]
OK, done!
You have to use the keypad as a 10 key (num lock on) and not as the special function keys.
So the numbers above the letter keys also work the same as long as you use the alt with them.
Several number keys were already using the ctrl modifier, none used the alt modifier so that is the one were using.

It's really interesting the available key bindings that can be used for any given.m key...
~~~
8 - up      357 forward -    127 - 56
2 - down    359 back -         0 - 50
4 - left    356 pitch up -    28 - 52
6 - right   358 pitch down -  30 - 54
9 - pgup    360 up -          57 - 57
3 - pgdown  361 down -        27 - 51
~~~
The 3xx are the special function key codes which I'm pretty sure I couldn't use.
The next column 0-127 are ctrl+the next column (50's).
Then the 50's are just the key code assigned to the strait key. Those are what I use with the alt modifier.
What is interesting is for example, if I understood what I was seeing correctly...
ctrl+8, could be assigned as 127 = (ctrl+8) or  as 56+mod-ctrl = (ctrl+8)

**Comment by gilbertohasnofb** [comment]
@wlbragg Cool, happy to hear you found the issue in the code! When the PR is ready, let me know and I will gladly review it :smile:

**Comment by wlbragg** [comment]
The seat portion is done and ready to be tested.
I still have the last part of the window de-fogging to finish.

**Comment by wlbragg** [comment]
I have a bug in the seat code I'm trying to isolate, so it isn't ready to test yet.

**Comment by wlbragg** [comment]
OK, the seat logic is finished. It's read for final testing.

**Comment by gilbertohasnofb** [comment]
@wlbragg I just tested it and I absolutely love this feature, such a great idea! I do have a couple of comments though:

* even though you can't move a seat sideways (which is correct), you should save the information of all axis to revert when pressing `q`. For instance, if you move the viewpoint sideways and then press `Q`, you are correctly reset to the default view. But if you move the viewpoint sideways and then press `q`, your view is reset only along certain axes, and you are left in the side position you were. Let me know if this is unclear, it's a bit difficult to explain by writing but very easy to see it in the sim.
* I do like that you used the `Alt` modifier instead of `Ctrl`, that's definitely better
* for some reason I can only move the view using the keypad buttons. Using the arrows or the PageUp and PageDown keys which are NOT in the keypad will not work. This might be a problem for people on laptops with small keyboards lacking the keypad.

**Comment by wlbragg** [comment]
I was still under the assumption I couldn't use the keypad portion, that you had to have numlock on. But that was because of a dumb mistake I had mad in the code. Then I though we still couldn't use it because it already has some of it mapped using the ctrl modifier for other actions. Since we are using the alt modifier instead now we can use the keypad without numlock on. I just pushed that change.
Still working on the left right view change resetting to center. I know exactly what you mean. I'm not sure that was ever programed that way? But I don't see any reason we can't, so I am working on it.

**Comment by gilbertohasnofb** [comment]
> Still working on the left right view change resetting to center. I know exactly what you mean. I'm not sure that was ever programed that way? But I don't see any reason we can't, so I am working on it.

Basically we just need to save the left and right information, but not add any shortcuts to alter it. This way we save all degrees of freedom so that `q` resets exactly to where the user has set it. Let me know once this gets done and I will re-test everything.

**Comment by wlbragg** [comment]
I decided that the "q" pilot reset view should be a quick reset to the saved pilot view no mater what view your currently in. "Q" is reset to the current view "default" as it always has been, including in the pilot view.

I hope everyone agrees with this concept. I look at it like the entire reason for "q" was to get back to your normal "user" pilot view. If your in the pilot view and you hit the "Q" then you immediately get the default pilot view. If you never change the from the default pilot view using the "alt + keys" and you hit "q" you will be taken to the "default" pilot view no mater what your current view is.
We never had a one key method to get back to the pilot view from anywhere, now we do.

I hope this all make sense.

**Comment by gilbertohasnofb** [comment]
Hmm, I am finding a weird bug, whenever I hit `q` the screen goes grey and I need to force close the sim. Maybe we don't need any fancy logic: `Q` takes you back to pilot view + default position, `q` takes you back to whatever view you were when you moved the seat + the position you saved.

**Comment by wlbragg** [comment]
Oops, clear your aircraft cache. I think that is all that is wrong. Some default info gets saved the first time this is used. You flag for that has already been set and I am now adding more saved data. In you case it is trying to rely on data that isn't there. Maybe as failsafe I should add that check, but it shouldn't ever be necessary if this code is used for the first time and doesn't change like we have been doing during the creation process.

**Comment by wlbragg** [comment]
I'm not saying there might not be other issues. But I have thoroughly been through this one and already hit it myself. 
That white blank screen is cause when the view repositioning code tries to use a property that doesn't exist.
This is the way I save the default user data the first time, after that if the flag is set then it never saves the original "default config" data again. I added more data to the code but your cache thought it already had saved all the original default data.
~~~
if (!getprop("sim/current-view/user-settings-saved")) {
            # initialize saved user seat settings for the first time
            setprop("sim/current-view/user/x-offset-m", getprop("sim/view/config/x-offset-m"));
            setprop("sim/current-view/user/y-offset-m", getprop("sim/view/config/y-offset-m"));
            setprop("sim/current-view/user/z-offset-m", getprop("sim/view/config/z-offset-m"));
            setprop("sim/current-view/user/default-field-of-view-deg", getprop("sim/view/config/default-field-of-view-deg"));
            setprop("sim/current-view/user/pitch-offset-deg", getprop("sim/view/config/pitch-offset-deg"));
            setprop("sim/current-view/user-settings-saved", 1);
        } else {
            #user defined saved seat position
            setprop("sim/current-view/x-offset-m", getprop("sim/current-view/user/x-offset-m"));
            setprop("sim/current-view/y-offset-m", getprop("sim/current-view/user/y-offset-m"));
            setprop("sim/current-view/z-offset-m", getprop("sim/current-view/user/z-offset-m"));
            setprop("sim/current-view/field-of-view", getprop("sim/current-view/user/default-field-of-view-deg"));
            setprop("sim/current-view/pitch-offset-deg", getprop("sim/current-view/user/pitch-offset-deg"));
        }
~~~
I could verify that the data I am using exists line by line instead of using a flag. This would eliminate the possibility of this happening whenever we make a change to this code.
~~~
if (getprop("sim/current-view/user/x-offset-m") == nil){
            setprop("sim/current-view/user/x-offset-m", getprop("sim/view/config/x-offset-m"));
        } else {
            setprop("sim/current-view/x-offset-m", getprop("sim/current-view/user/x-offset-m"));
        }
        if (getprop("sim/current-view/user/y-offset-m") == nil){
            setprop("sim/current-view/user/y-offset-m", getprop("sim/view/config/y-offset-m"));
        } else {
            setprop("sim/current-view/y-offset-m", getprop("sim/current-view/user/y-offset-m"));
        }
        if (getprop("sim/current-view/user/z-offset-m") == nil){
            setprop("sim/current-view/user/z-offset-m", getprop("sim/view/config/z-offset-m"));
        } else {
            setprop("sim/current-view/z-offset-m", getprop("sim/current-view/user/z-offset-m"));
        }
        if (getprop("sim/current-view/user/default-field-of-view-deg") == nil){
            setprop("sim/current-view/user/default-field-of-view-deg", getprop("sim/view/config/default-field-of-view-deg"));
        } else {
            setprop("sim/current-view/field-of-view", getprop("sim/current-view/user/default-field-of-view-deg"));
        }
        if (getprop("sim/current-view/user/pitch-offset-deg") == nil){
            setprop("sim/current-view/user/pitch-offset-deg", getprop("sim/view/config/pitch-offset-deg"));
        } else {
            setprop("sim/current-view/pitch-offset-deg", getprop("sim/current-view/user/pitch-offset-deg"));
        }
~~~

**Comment by wlbragg** [comment]
I actually think I like the second approach better and it is less likely to screw up. I'm going to try it that way and delete some of the cached data but not all and see if it works.

**Comment by wlbragg** [comment]
That worked well, now what happened to you can't happen, ever and it removed the need for a new property.

**Comment by gilbertohasnofb** [comment]
Emptying the cache did the trick, but I found another bug, here is how to reproduce it:

* start in the default view
* now move the seat anywhere using the Alt plus keys, say raise it a bit
* at this point, pressing `Q` resets back to the default view and pressing `q` goes back to your setting
* cool, so after pressing `q`, move the seat a bit more upwards. Press `Q` and `q` and confirm that the user's prefered seat position has been updated. So far so good. Let's call this seat position "X"
* now press `Q`
* move the seat using the Alt keys, say lower it a bit
* now if you press `q`, it will take you back to the position X instead. It is as if there are two modes, one enabled after pressing `Q` and which will ignore any changes you make and one enabled after `q` which updates the user's preferred seating position.

Let me know if you have difficulties reproducing this one, again it's tricky with text but easy in the sim.

**Comment by wlbragg** [comment]
@gilbertohasnofb I found the issue, good catch. But i have a question on how exactly to proceed.

What is happening is when you use alt+movement-key you add or subtract .005 from the previous saved user property axis. However if you have moved using the Q key you are starting from a new position but are still adding or subtraction .005 to the original user saved setting for that axis, instead of starting from the new position you now in.
OK, easy fix, whenever you hit the alt+movement-key I update that user position axis to the current view axis, problem solved. Only thing is, that doesn't account for any other axis, they all need to be updated every time you move with the alt key down.
So I guess that is what I am going to do.
Bottom line now, when ever you hold alt down and move position it starts a new clean slate and saves from that starting position + any subsequent movement.

**Comment by wlbragg** [comment]
@gilbertohasnofb OK this is fixed and pushed.

**Comment by gilbertohasnofb** [comment]
@wlbragg works perfectly now, thanks for the hard work! :clap:

**Comment by gilbertohasnofb** [comment]
such a cool feature, really

**Comment by legoboyvdlp** [comment]
It worked well when I tried it!


However, I thought the rate of fog / frosting might have been too fast - it only took about 30 seconds to get maximum frosting at -2 degC? When I applied max heat and 10% airflow it went very rapidly to full fogging.
